### PR TITLE
UCT/IB: Improve DC full handshake logic

### PR DESCRIPTION
**Pull request by gleon99**
## What
1. Use address version flag.
2. Check whether HW reports better performance when no RDMA_WRITE

## Why ?
1. Currently device flag is used, which is a bug (incorrect value).
2. Only set `rdma_wr_disabled` bit when it has an impact.

### CODE REVIEW

**Review by yosefe — DISMISSED** [review]

File: src/uct/ib/mlx5/dv/ib_mlx5_ifc.h [hunk 1]
```diff
@@ -272,7 +272,8 @@ struct uct_ib_mlx5_cmd_hca_cap_bits {
     uint8_t    rc[0x1];
 
     uint8_t    uar_4k[0x1];
-    uint8_t    reserved_at_241[0x9];
+    uint8_t    dci_no_rdma_wr_optimized_performance[0x1];
+    uint8_t    reserved_at_241[0x8];
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5_ifc.h:276]: should we rename it to reserved_at_242?

File: src/uct/ib/dc/dc_mlx5.c [hunk 2]
```diff
@@ -324,7 +324,8 @@ static ucs_status_t uct_dc_mlx5_iface_create_dci(uct_dc_mlx5_iface_t *iface,
         attr.super.max_inl_cqe[UCT_IB_DIR_RX] = 0;
         attr.uidx             = htonl(dci_index) >> UCT_IB_UIDX_SHIFT;
         attr.full_handshake   = full_handshake;
-        attr.rdma_wr_disabled = iface->flags & UCT_DC_MLX5_IFACE_FLAG_DISABLE_PUT;
+        attr.rdma_wr_disabled = (iface->flags & UCT_DC_MLX5_IFACE_FLAG_DISABLE_PUT) &
+                                (md->flags & UCT_IB_MLX5_MD_FLAG_NO_RDMA_WR_OPTIMIZED);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:328]: maybe add ucs_debug message to uct_ib_mlx5_devx_create_qp, that indicates which value we passed to rdma_wr_disabled

File: src/uct/ib/mlx5/dv/ib_mlx5_ifc.h (hunk 1 above)
Comment (gleon99) [reply src/uct/ib/mlx5/dv/ib_mlx5_ifc.h:276]: Nice!

File: src/uct/ib/dc/dc_mlx5.c (hunk 2 above)
Comment (gleon99) [reply src/uct/ib/dc/dc_mlx5.c:328]: Are you sure? In general I don't see such messages for other flags and fields in this file or in `ib_mlx5_dv.c`

File: src/uct/ib/mlx5/dv/ib_mlx5_dv.c [hunk 3]
```diff
@@ -157,6 +157,8 @@ ucs_status_t uct_ib_mlx5_devx_create_qp(uct_ib_iface_t *iface,
 
     qp->devx.obj = mlx5dv_devx_obj_create(dev->ibv_context, in, sizeof(in),
                                           out, sizeof(out));
+    ucs_debug("created DevX DCI 0x%x, rdma_wr_disabled=%d", qp->devx.obj,
```
Comment (yosefe) [inline src/uct/ib/mlx5/dv/ib_mlx5_dv.c:160]: it should be after creating the object successfully, and "DCI 0x%x" should print the qp_num
Also sorry but it should be in uct_dc_mlx5_iface_create_dci, not uct_ib_mlx5_devx_create_qp

**Review by yosefe — APPROVED** [review]

**Review by yosefe — APPROVED** [review]

**Pull request merged**